lib/datadog/appsec/waf.rb

```diff
@@ -482,6 +479,33 @@ class Handle
 
         attr_reader :ruleset_info
 
+        def self.update(handle, data)
+          data_obj = Datadog::AppSec::WAF.ruby_to_object(data, coerce: false)
+          ruleset_info = LibDDWAF::RuleSetInfo.new
+          new_handle = Datadog::AppSec::WAF::LibDDWAF.ddwaf_update(handle.handle_obj, data_obj, ruleset_info)
+
+          return if new_handle.null?
```

Comment:
This is inconsistent with `initialize`, which raises an exception on failure. I'd rather keep a `nil` return value to mean "there was no update".

Reply:
The behaviour of the API is designed to consistently raise exceptions on unhappy paths, and this would make a special case:

- if the update does not fail but does nothing it returns `nil`: this is a happy path
- if the update failed it returns `nil`: this is an unhappy path

Therefore the caller can't know whether the update fails or did not fail, and cannot e.g produce an error log on the unhappy path, otherwise we'd also log the happy path constantly.

IIRC this is a limitation of `ddwaf_update` which returns `null` in both cases. I think @Anilm3 is aware of that and is working on improving that situation†.

That said, at this level we can work around that a bit and introspect `ruleset_info` for errors, and raise if any. I'd rather not have the API consumer (ddtrace) have to introspect `ruleset_info` for errors, as it is largely opaque to them. `ruleset_info` is limited as it only reports errors for rules, but I think that would be the most frequent ones and that's better than nothing.

† There was discussion about what a NOOP update would do, e.g return the previous handle as is.

Reply:
I understand. Thanks for the detailed explanation.

I agree with the statement: We should consistently raise exceptions if an error occurs. 

> That said, at this level we can work around that a bit and introspect ruleset_info for errors, and raise if any. I'd rather not have the API consumer (ddtrace) have to introspect ruleset_info for errors, as it is largely opaque to them. ruleset_info is limited as it only reports errors for rules, but I think that would be the most frequent ones and that's better than nothing.


That would be a good mid-solution until we have waf returning either the same handle for NOOP or a NULL pointer for a failed update. 

Unfortunately, the code path in the waf, that updates the info struct, only reports changes when passing the `rules` key

https://github.com/DataDog/libddwaf/blob/15b228207547d9d897b2a85b6a93cf22a1261e46/src/ruleset_builder.cpp#L65

https://github.com/DataDog/libddwaf/blob/15b228207547d9d897b2a85b6a93cf22a1261e46/src/ruleset_builder.cpp#L203

https://github.com/DataDog/libddwaf/blob/fe0bd8389ba1dffeb0579363f09d43a5dc803fa2/src/parser/parser_v2.cpp#L407

https://github.com/DataDog/libddwaf/blob/fe0bd8389ba1dffeb0579363f09d43a5dc803fa2/src/parser/parser_v2.cpp#L413

https://github.com/DataDog/libddwaf/blob/fe0bd8389ba1dffeb0579363f09d43a5dc803fa2/src/parser/parser_v2.cpp#L417

https://github.com/DataDog/libddwaf/blob/fe0bd8389ba1dffeb0579363f09d43a5dc803fa2/src/parser/parser_v2.cpp#L420


So I think if we want to be consistent with the initialize API we can check if the new handle pointer is NULL and so `fail`

Or we can keep it as returning `nil` and make sure to update the API once the `waf` provide us with better error information